### PR TITLE
WT-8372 format mirror objects

### PR DESCRIPTION
**Pull request by keithbostic**
This is a set of changes to add mirrored objects to format, that is, when there are multiple tables, format randomly picks a subset of them and mirrors content between them. Normal verify of the tables is enhanced so that after each individual object is verified, mirrored objects are compared to each other, asserting if they contain different content. The changes are relatively isolated, but there's one big change where the primary format operations loop was split into two parts, the table, transaction and operation selection loop, and the per-table code to perform a single operation.

I've requested a lot of reviewers (I'm listing the same reviewers as I did for WT-3445, since if you dug into that change, well, this is more of the same), plus a few others. Please don't hesitate to remove yourself from the review list, or limit your review to specific areas you're interested in (or to areas where I've called you out below with a request for your review).

Notable changes I think people might disagree with:

1. Previously, format tried to dump the cursor's pages if it ever found an unexpected key/value pair, based on the `g.page_dump_cursor` value when it exited. After discussions with @keitharnoldsmith, I've hooked that work into the WiredTiger library itself: there's a field in the `WT_SESSION_IMPL` structure, and it's set to a callback function called from `__wt_btcur_search()`. I've resisted this change for a long time, because connecting test programs to the library seems ugly, but I know that @kommiharibabu and @sueloverso wanted to decrease the chance of format simply dropping core when dumping pages, and this should minimize the chances of getting a core dump instead of useful information when format fails. @sueloverso, @kommiharibabu, would you two please review that change? It's in commit 1973cb6.

2. I've removed all of the cursor order checking in format, that is, when moving a cursor around in the table, it doesn't check that rows are returned in the correct order. I did this because it made another change easier, but also because when the library build is `HAVE_DIAGNOSTIC`, the library does stronger order checking that format ever did, and it seemed reasonable to let the library own that problem.

3. In `WT_CURSOR.modify` calls, if there's a chunk of data that's not being set by the modify call, the bytes are set to a default, nul byte. That makes debugging format more difficult that it needs to be because that's the only key/value characters that aren't printable. Further, when tracking operations in format, there's a lot of wasted effort converting those strings to printable records. I've added a hack that allows test programs to set the modify pad byte, mostly because I'm just tired of seeing escaped nul bytes in format output. @quchenhao, that change is in commit 15c3c63 and I think you're the most familiar of us with that area, would you please take a look?

4. Previously, there was a `-T` option to format that configured how tracing is done. I removed the `-T` option in favor of finer granularity of tracing by using configuration options. I don't think the `-T` option is widely used, so I don't think this will be a backward compatibility problem for users, and the new configuration options will be ignored by any previous versions of format.

Finally, @lukech, there's a change to `format.sh` in 8edd94f that replaces the calls to the `wt` utility for abort/recovery testing with calls to format, using a new `-v` option that, when combined with the existing `-R` option opens the databases, verifies them, and then exits. That's because the format utility knows how to compare mirror groups against each other, and `wt` doesn't. @lukech, would you please review that change?

UPDATE:
We merged the first set FLCS changes a couple of days ago, so I went ahead and included the changes to once again support FLCS in format (FLCS in format hadn't been updated in a long time because the WiredTiger library hasn't supported FLCS in years, and there had been extensive bit rot). This branch currently supports FLCS to some extent, although it hasn't been stress tested at all.

### CODE REVIEW

**Comment by keithbostic** [comment]
> Sorry, I mean FORMAT_PAD_BYTE. Can we just directly use FORMAT_PAD_BYTE for the place that we use `modify_pad_byte `?

@quchenhao, doesn't that require doing something like:
```
    if (value->size <= offset) {
        if (value->size < offset)
            memset((uint8_t *)value->data + value->size,
#ifdef STANDALONE_FORMAT_BUILD
              sformat ? ' ' : FORMAT_PAD_BYTE, 
#else         
              sformat ? ' ' : 0, 
#endif        
              offset - value->size);
```
in `src/support/modify.c: __modify_apply_one()`?

Am I missing something, or do you prefer that approach for some reason?

File: test/format/format.h [hunk 1]
```diff
@@ -309,6 +318,7 @@ typedef struct {
     uint64_t commit; /* operation counts */
     uint64_t insert;
     uint64_t prepare;
+    uint64_t modify;
```
Comment (sueloverso) [inline test/format/format.h:321]: alphabetize

File: test/format/README [hunk 2]
```diff
@@ -0,0 +1,62 @@
+Tracing format operations can be useful in debugging.
+
+Turn on tracing with the "-t" flag:
+
+    -t	Trace operations into a WiredTiger log
+
+That causes format to trace operations in a WiredTiger log.
```
Comment (sueloverso) [inline test/format/README:7]: Not sure who the audience is but when I read that I was thinking this was using `log_printf` into the actual WiredTiger log, i.e. `WiredTigerLog.000000#`.

**Review by sueloverso — COMMENTED** [review]
I have made a few comments, but need to take another pass at this.

File: test/format/config.c [hunk 3]
```diff
@@ -351,7 +350,8 @@ config_run(void)
     if (g.backward_compatible)
         config_backward_compatible();
 
-    config_cache(); /* Cache */
+    config_mirrors(); /* Mirrors */
+    config_cache();   /* Cache */
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply test/format/config.c]: Comment added in c083600c3.

File: test/format/config.c [hunk 4]
```diff
@@ -935,6 +941,101 @@ config_lsm_reset(TABLE *table)
...
+        if (tables[i]->type != FIX)
+            break;
+    if (ntables < 2 || i > ntables) {
+        if (explicit_mirror)
+            WARN("%s", "table selection didn't support mirroring, turning off mirroring");
+        return;
+    }
+
+    /*
+     * Don't mirror if there are custom collator, it would complicate the cursor traversal when
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply test/format/config.c]: Fixed in aa9ce0600.

File: test/format/format.h (hunk 1 above)
Comment (keithbostic) [reply test/format/format.h:320]: Fixed in 2e4315dfc.

File: test/format/README (hunk 2 above)
Comment (keithbostic) [reply test/format/README]: That's exactly what it does, it uses `log_printf` to insert messages into the WiredTiger log files.

I clarified a bit in 71ba2cef7.

Comment (sueloverso) [reply test/format/README]: And that is exactly why I need to take another pass. I missed that! Thanks for the clarifications in the readme.

**Comment by tammybailey** [comment]
I'm removing myself from the reviewers because I don't have time to properly review the change in its entirety. Happy to help with a smaller chunk if anyone is overwhelmed though - let me know.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM.

I've focused on:

- The hook to validate repeated reads
- Looking at the code that mirrors operations and the code that verifies mirrors to make sure they match up.

I've left a few comments/suggestions, but only a possible comment error seems like a required change.

**Comment by keithbostic** [comment]
This branch fails in a couple of different scenarios, but most importantly, it fails abort testing. There's no point in doing further review until at least that issue is resolved, so I'm going to close this branch down for now.

File: test/format/README [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    trace.read=1	trace read operations
+    trace.timestamp=1	trace timestamp operations (setting oldest, stable TS)
+
+By default, operations are logged into a separate database from the run, "OPS.TRACE" in the
+run's home directory. If you want to trace format operations with logs from in the WiredTiger
+library, you can log format operations into the primary database:
+
+    trace.local=1	trace into the primary database
+
+By default, only 10 log files are retained at a time, but you can increase that number if you
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply test/format/README:24]: > Should this be 25 log files?

Why 25 log files?

File: test/format/ops.c [hunk 6]
```diff
@@ -980,80 +1075,97 @@ ops(void *arg)
...
+            tinfo->table = table;
+            ret = table_op(tinfo, intxn, iso_level, op);
+            testutil_assert(ret == 0 || ret == WT_ROLLBACK);
+            skip2 = table;
+        }
+        if (ret == 0 && table->mirror)
+            for (i = 1; i <= ntables; ++i)
+                if (tables[i] != skip1 && tables[i] != skip2) {
+                    tinfo->table = tables[i];
+                    ret = table_op(tinfo, intxn, iso_level, op);
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply test/format/ops.c:1138]: I thought a little bit about exactly that, but eventually decided that any divergence should be caught in a future mirror check (assuming the divergence doesn't come back "together", which should be unlikely). Is there a path where a future mirror check won't find the problem?

**Pull request closed without merging**